### ISSUE TITLE
Problem with PBKDF2 Hash

### ISSUE THREAD

**Issue by paul1982** [iegomez/mosquitto-go-auth#173]
The plugin is working correct now..
But I still have problems with the PBKD2 hash.

This is what i have done:
My test account has the password 5633ac. (so my client sent this string as password.)
On my server I have generate a password with the command np. it ask for the password and I type 5633ac.
It generate every time a different password, I expected to be the same everytime.
If I put the generated hash password in my database it is not accepted by the plugin.

So I added a couple of debug rules in the source. and notched that the plugin generate a different password. (if i added that password in my db the plugin accept the user)

So there is something going wrong with the hashing.
The settings I use:

auth_opt_hasher pbkdf2
auth_opt_hasher_salt_size 16           # salt bytes length
auth_opt_hasher_iterations 100000      # number of iterations
auth_opt_hasher_keylen 64              # key length
auth_opt_hasher_algorithm sha512       # hashing algorithm, either sha512 (default) or sha256
auth_opt_hasher_salt_encoding          # salt encoding, either base64 (default) or utf-8

I can not find what is used as salt string.. I think I don't understand the concept..

I hope someone can push me in the right direction.

**Comment by iegomez** [comment]
> On my server I have generate a password with the command np.

Please use the plugin provided `pw` utility to generate the password hash, e.g. using defaults:
```
./pw -p 5633ac
```

Since `pw` and the plugin use PKBDF2 by default, as well as the same same default hashing options when not provided, you can omit all `auth_opt_hasher` options and it should work out of the box with the password generated in the previous step.

> It generate every time a different password, I expected to be the same everytime.

Yes, that's because it uses a random salt at the moment of generating the hash. This salt is encoded in the hash itself so when checking the password the plugin uses that same salt to regenerate the hash before comparing. There's no user action needed regarding the salt unless you want to customize its size and/or encoding to match those used by external parties.

TL;DR: generate the password with `pw` with no flags other than `-p` for passing the password, and omit all hashing options in the plugin's configuration, and that should work.

**Comment by paul1982** [comment]
Oke, now it works.... 
So I think i used the wrong password tool.

Sorry for the stupid mistake.

I want to thank you for you quick response and good job on the plugin. It fills a big gab in the mosquitto project.
Is there some kind of support button where I can make a little donation?

**Comment by iegomez** [comment]
I appreciate it, @paul1982, but this is just a hobby project for me, maybe it just went a bit too far. 😅 
So no need to donate, I'm doing well in my day job, just keep using it and report any issues so we can make it better!

Cheers!
